src/rgw/rgw_sal_rados.cc

```diff
@@ -2943,40 +2992,72 @@ int RadosZoneGroup::get_placement_tier(const rgw_placement_rule& rule,
   return 0;
 }
 
-int RadosZone::get_zonegroup(const std::string& id, std::unique_ptr<ZoneGroup>* zonegroup)
+int RadosZoneGroup::get_zone_by_id(const std::string& id, std::unique_ptr<Zone>* zone)
 {
-  ZoneGroup* zg;
-  RGWZoneGroup rzg;
-  int r = store->svc()->zone->get_zonegroup(id, rzg);
-  if (r < 0)
-    return r;
+  RGWZone* rz = store->svc()->zone->find_zone(id);
+  if (!rz)
+    return -ENOENT;
 
-  zg = new RadosZoneGroup(store, rzg);
-  if (!zg)
-    return -ENOMEM;
+  Zone* z = new RadosZone(store, clone(), *rz);
+  zone->reset(z);
+  return 0;
+}
 
-  zonegroup->reset(zg);
+int RadosZoneGroup::get_zone_by_name(const std::string& name, std::unique_ptr<Zone>* zone)
+{
+  rgw_zone_id id;
+  int ret = store->svc()->zone->find_zone_id_by_name(name, &id);
+  if (ret < 0)
+    return ret;
+
+  RGWZone* rz = store->svc()->zone->find_zone(id.id);
+  if (!rz)
+    return -ENOENT;
+
+  Zone* z = new RadosZone(store, clone(), *rz);
+  zone->reset(z);
   return 0;
 }
 
-const rgw_zone_id& RadosZone::get_id()
+std::unique_ptr<Zone> RadosZone::clone()
 {
-  return store->svc()->zone->zone_id();
+  if (local_zone)
+    return std::make_unique<RadosZone>(store, group->clone());
+
+  return std::make_unique<RadosZone>(store, group->clone(), rgw_zone);
+}
+
+const std::string& RadosZone::get_id()
+{
+  if (local_zone)
+    return store->svc()->zone->zone_id().id;
+
+  return rgw_zone.id;
```

Comment:
i'm having trouble following this `local_zone` flag. what does it mean, and in what cases would the `rgw_zone` member variable be different from what we'd get through `store->svc()->zone->`?

Reply:
The problem is that "zone" is overloaded in the rados code.  There's both "zone" as an object that represents a zone within a zonegroup, and "zone" as an interface that gets information about the "current" zone.  "local_zone" represents the interface.

The issue is that the "interface" version doesn't really have a RGWZone that we can use to represent it.  I'd like to have a better way of doing this, but I'm not sure what it would look like.